### ISSUE TITLE
Compile instruction with gnu make and gcc

### ISSUE THREAD

**Issue by ismail** [nidud/asmc#8]
First of all, thanks for this! In case you didn't notice it made 7-zip on Linux a reality: https://sourceforge.net/p/sevenzip/discussion/45797/thread/7fe6c21efa/

My question is, how does one compile this using gcc and gnu make, is that at all possible? I see Makefile references to Open Watcom compiler, but I'm afraid access to that compiler is not really mainstream for Linux users.

Thanks in advance!

**Comment by nidud** [comment]
I guess it's possible but I plan on cleaning up the source in the near future, separate the C source from the assembler implementation and remove the dependency between them.

This will simplify making more portable versions but I need to write a few more ASM source files first.

**Comment by ismail** [comment]
Thanks, this would be very much appreciated, this way we can cleanly package 7-zip for Linux distributions and make asmc available easily.

**Comment by teoberi** [comment]
Where did the Linux version of ASMC disappear?

**Comment by nidud** [comment]
Well, nothing really goes away here. As there haven't been any changes added to the Linux (or C++) version for some time the latest version is here:
https://github.com/nidud/asmc/tree/0d32752205547ec3f5c995e6005f282ea7d400bf

As mention above the C++ source will be separated and a new repository is added for this.

https://github.com/nidud/asmcpp

However, it still depends on one external library for the quad float implementation.

**Comment by ismail** [comment]
> As mention above the C++ source will be separated and a new repository is added for this.
> 
> https://github.com/nidud/asmcpp

This looks good, but it needs a Linux compatible definitions for memicmp/stricmp functions.

> 
> However, it still depends on one external library for the quad float implementation.

GCC ships a quadmath library by default.

**Comment by teoberi** [comment]
> My question is, how does one compile this using gcc and gnu make, is that at all possible?

I am also interested in compiling asmcpp in Linux.

**Comment by nidud** [comment]
> This looks good, but it needs a Linux compatible definitions for memicmp/stricmp functions.

Yes, there have to be some adjustment.

> GCC ships a quadmath library by default.

Not really a portable solution as you still have to add most of the code needed. The work size internally use a 128 bit mantissa which is scaled down to REAL16, 10, 8, 4, and REAL2 (in addition to __int128 support), so this needs to be converted to C++ eventually.

**Comment by nidud** [comment]
I removed the asmcpp repository and moved the source back here. The problem is that you need external assembly source to build Asmc as the complexity exceeds the capability of most C compilers at this point. 

It is obviously possible to do all of this also in C (as least in 64-bit), but the amount of code needed and external libraries added (which needs to be tailored for each compiler and bitness) would make this a rather complicated task. So to build Asmc in Linux you basically need to install Asmc first or supply the libraries needed.

**Comment by darealshinji** [comment]
Can you write a Makefile to compile the assembler on Linux? I have written one but get errors like these:
```
$ make -f Makefile.gcc
./linux/bin/asmc -q -Zp4 -coff -Cs -I./src/h -I../../include src/asmc.asm
../../include/winnt.inc(2339) : error A2008: syntax error : RECORD
 ../../include/winnt.inc(2339): Included by
  ../../include/signal.inc(3): Included by
   src/asmc.asm(9): Main line code
```

**Comment by nidud** [comment]
>atofloat.c:(.text+0x16e): undefined reference to `__cvth_q'

I added quadmath.lib to the Linux source so you may try adding this and see how that goes.

**Comment by darealshinji** [comment]
That works. But the issue remains that the assembler isn't bootstrapped completely from source.
Can you add the x86-64 quadmath library too? I'll disassemble it to GNU assembler so that we have GCC compatible source code of everything.

**Comment by darealshinji** [comment]
Update: I've assembled the x86_64 sources and disassembled them to GNU asm files. Now I can compile asmc as 32 bit and 64 bit from source. I will make a pull request later.

**Comment by darealshinji** [comment]
Here are the changes I had to do to build asmc with GCC and GNU assembler: https://github.com/darealshinji/asmc/commit/a356ac1e419c053c5761463d9312c6bd48dd69c4

But it still doesn't assemble all the 7zip asm files correctly...

**Comment by nidud** [comment]
Did you test the version built with the library?

The stdio functions use quad float so the source is embedded in LIBC. If you extract the sources I think most of them could easily be converted to C but you still left with a set of core assembly functions.

**Comment by darealshinji** [comment]
The 32 bit asmc build works without issues, using either the GNU asm files or the quadmath.lib library.
The 64 bit build chokes on one file:
```
 Assembling: LzmaDecOpt.asm
ASMC : fatal error A1901: Internal Assembler Error
```

**Comment by darealshinji** [comment]
Can [GCC libquadmath](https://gcc.gnu.org/onlinedocs/libquadmath/) or [Intel intrinsics](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#) be used to replace some of the asm code?

**Comment by nidud** [comment]
I guess they could probably replace all of them.

**Comment by nidud** [comment]
> The 64 bit build chokes on one file:

I haven't built a functional 64 bit version for Windows either, so it's not necessarily the quadmath functions that is the reason for this.

**Comment by teoberi** [comment]
Indeed, 32-bit Asmc can be compiled, but:
`ls -hl bin/asmc `

> -rwxr-xr-x 1 sysadmin sysadmin 397K dec  5 19:42 bin/asmc

`ls -hl asmc_gcc` 

> -rwxrwxr-x 1 sysadmin sysadmin 599K dec  5 19:59 asmc_gcc

**Comment by darealshinji** [comment]
The GCC build is likely bigger because it's built from C sources with speed optimizations and not size optimizations.

**Comment by nidud** [comment]
It's possible to build a Linux version from the assembly source now so I added binaries for v2.33.19 to the [linux](../tree/master/source/asmc/linux) directory for testing. Note that ASMC64 is similar to ML64 as it do not understand 32-bit directives. The default for the Linux version is ELF64 so the command line option -elf64 is not needed.

The C version is 2.32.55 so if this works it will be an upgrade but the problem remains: you need Asmc to build Asmc as GCC needs GCC.

**Comment by teoberi** [comment]
I'm trying to compile 7-Zip 21.06 using asmc/source/asmc/linux/asmc-2.33.19/asmc64 but I have a problem!
`make -f ../../cmpl_gcc_x64.mak`

> asmc64 -elf64 -DABI_LINUX -Fob/g_x64/ ../../../../Asm/x86/LzmaDecOpt.asm
> Asmc Macro Assembler (x64) Version 2.33.19
> Copyright (C) The Asmc Contributors. All Rights Reserved.
> 
>  Assembling: ../../../../Asm/x86/LzmaDecOpt.asm
> ASMC : error A2005: symbol redefinition : _LINUX
> ASMC : fatal error A1901: Internal Assembler Error
> make: *** [../../7zip_gcc.mak:1157: b/g_x64/LzmaDecOpt.o] Segmentation fault (core dumped)
> make: *** Deleting file 'b/g_x64/LzmaDecOpt.o'

**Comment by nidud** [comment]
I downloaded the source and ASMC64 failed if the -elf64 argument was used. This is fixed in version 2.33.20.

However, the new version is more Masm compatible with regards to error messages so this will fail (7zAsm.asm):
```
ifdef RAX
  x64 equ 1
endif
``` 
Masm will reject this as IF[n]DEF expects a plain symbol as argument. Asmc defines [various symbols](../tree/master/doc/symbol/readme.md) for this, like _WIN64 or _LINUX, but a more portable way will be to use the text macro @wordsize:
```
if @wordsize eq 8
  x64 equ 1
endif
```

**Comment by teoberi** [comment]
Thanks @nidud!
ASMC64 version 2.33.20 solved the problem.
I also made the change to 7zAsm.asm which means:
ifdef RAX -> if @wordsize eq 8
Now I can compile 7-Zip with a single warning at AesOpt.asm -> ifdef ymm0

**Comment by teoberi** [comment]
@nidud we have a problem with the change to 7zAsm.asm
Igor Pavlov says [here](https://sourceforge.net/p/sevenzip/discussion/45797/thread/b463f34259/?page=2#a277/07a0) that:

> `if @wordsize eq 8`
> doesn't work with MASM.

Is there another solution to this problem?

**Comment by nidud** [comment]
Well, let me rant a bit around the subject.

The problem is that the IF[n]DEF directive should only be used for plain symbols, so used with reserved words like RAX or DWORD is erroneous.

Example:
```
ifdef DWORD
echo <DWORD defined>
else
echo <DWORD not defined>
endif
```
This is an unreported error in Masm which always return 0. The same occur in IFDEF RAX, which for ML is just an undefined symbol (same for JWASM/ASMC), but for ML64 an (unreported) error which return true.

So this becomes a logistical problem as the same assembler is used for both 32 and 64-bit code (ASMC) but Masm use two assemblers: ML do not understand 64-bit, and ML64  do not understand 32-bit.

`IFNDEF @cpu ; ML(1), ML64(0), ASMC(1)`

My take is that these things should be resolved in the MAKEFILE as you there have to decide the bitness (use ML or ML64, -elf or -elf64) by adding command line options like -Dx64 and -Duse_vaes_256.

In Asmc you will normally add -arch:AVX and do:
```
ifdef __AVX__
  use_vaes_256 equ 1
```
However, this should work:
```
ifdef __JWASM__ ; Asmc/Uasm/JWasm
if @wordsize eq 8
 x64 equ 1
endif
else ; Masm
ifndef @cpu
 x64 equ 1
endif
endif
```
The IFDEF YMM0 could then be replaced with IFDEF x64

**Comment by nidud** [comment]
This will also work:
```
ifdef @wordsize
if @wordsize eq 8
 x64 equ 1
endif
else
 x64 equ 1
endif

```

**Comment by nidud** [comment]
The C source is now removed and the Linux binaries are in the bin directory.

I fixed a few bugs and rewrote some of the header files to be more "Linux-friendly" so it's possible to build Asmc using GCC now. I also added the quadmath and crypt source so no external libraries is needed.

The i386 libraries appear to be missing in the 64-bit Mint that I'm using, but the 32-bit version was able to build from the GCC32.mak file.

**Comment by teoberi** [comment]
Hi @nidud!
The i386 libraries can be installed in the 64-bit Mint 20 (Mint is based on Debian and Ubuntu). In Ubuntu (what I use) at least that libraries can be installed.
Why did you name the make file GCC32.mak? It's a little confusing.
So to build asmc64 in Linux with GCC we need asmc (line 158 in GCC32.mak).
I am trying to build asmc64 in Linux with the following steps:
1. `chmod 744 asmc/bin/asmc`
2. `cd asmc/source/asmc/`
3. `make -f GCC32.mak`

The result is:
> ../../bin/asmc -nologo -nolib -Zp4 -elf -Cs -I../../include -Isrc/inc src/asmc.asm
>  Assembling: src/asmc.asm
> ../../bin/asmc -nologo -nolib -Zp4 -elf -Cs -I../../include -Isrc/inc src/asmerr.asm
>  Assembling: src/asmerr.asm
> ../../bin/asmc -nologo -nolib -Zp4 -elf -Cs -I../../include -Isrc/inc src/assemble.asm
>  Assembling: src/assemble.asm
> ../../include/winerror.inc(5) : fatal error A1000: cannot open file : specstrings.inc
> ASMC : fatal error A1901: Internal Assembler Error
> make: *** [GCC32.mak:158: src/assemble.o] Segmentation fault (core dumped)
>

**Comment by darealshinji** [comment]
> The C source is now removed and the Linux binaries are in the bin directory.

The question now would be why to build from source when a binary is available.

By the way the error above can be fixed by renaming "SpecStrings.inc" to "specstrings.inc" (Linux is case-sensitive), But then I get this:
```
../../bin/asmc -nologo -nolib -Zp4 -elf -Cs -I../../include -Isrc/inc src/asmc.asm
 Assembling: src/asmc.asm
../../bin/asmc -nologo -nolib -Zp4 -elf -Cs -I../../include -Isrc/inc src/asmerr.asm
 Assembling: src/asmerr.asm
../../bin/asmc -nologo -nolib -Zp4 -elf -Cs -I../../include -Isrc/inc src/assemble.asm
 Assembling: src/assemble.asm
../../include/wingdi.inc(65) : error A2005: symbol redefinition : ERROR
 ../../include/wingdi.inc(65): Included by
  ../../include/windows.inc(447): Included by
   ../../include/rpc.inc(5): Included by
    ../../include/segment.inc(8): Included by
     src/assemble.asm(13): Main line code
src/assemble.asm(786) : error A2008: syntax error : SegmentInit
make: *** [GCC32.mak:158: src/assemble.o] Error 1
```

**Comment by teoberi** [comment]
> The question now would be why to build from source when a binary is available.

For my servers I use the basic skeleton from Slackware64 but I compile all the necessary packages from the sources.

I also got the SegmentInit error but now I don't have time to dig.

**Comment by nidud** [comment]
Files like win*.inc and specstrings.inc should not be included, so this is strange.

**Comment by nidud** [comment]
I downloaded the zip file and it assembles without errors in 32 and 64  bit. It builds asmc and asmc64 in the 32-bit verion.
```
make -f GCC32.mak
make -f GCC32.mak X64=1
```
Could be some old include files?

> Why did you name the make file GCC32.mak? It's a little confusing.

The make file obviously need some work but this works for the 32-bit version of GCC, hence the name. It's more a test to see if it is capable of reproducing itself in the Linux environment, or a minimal test to see if it actually works.

**Comment by nidud** [comment]
>  libraries can be installed.

Yes. I installed the libraries so now it works. 

    sudo apt-get install gcc-multilib

I also cleaned up the make file a bit and renamed it to [GccLinux](https://github.com/nidud/asmc/blob/master/source/asmc/GccLinux.mak).

**Comment by jloqfjgk** [comment]
Is it correct that I need to use the prebuilt (1st stage / bootstrap?) binary `bin/asmc` to compile asmc?
If yes, is there any way to compile that 1st stage binary from source?

(I am wondering about that because I guess distro package maintainers may want to build the program from scratch instead of using a supplied binary)

**Comment by teoberi** [comment]
> Is it correct that I need to use the prebuilt (1st stage / bootstrap?) binary `bin/asmc` to compile asmc?

Yes, now to build asmc64 in Linux with GCC we need asmc (line 89 in GccLinux.mak)
 

> If yes, is there any way to compile that 1st stage binary from source?

From what I understand from @nidud now, at this stage NO.

> ... but the problem remains: you need Asmc to build Asmc as GCC needs GCC.

> 
> (I am wondering about that because I guess distro package maintainers may want to build the program from scratch instead of using a supplied binary)

You're absolutely right, Linux distribution maintainers avoid binaries and prefer to compile from sources (the Linux way), although there are exceptions such as commercial applications that sometimes need to be installed in Linux (eg some versions of antivirus, in my case Sophos for Linux).
The old asmc was generated using Open Watcom in Windows for Linux (OWLinux.mak).

**Comment by nidud** [comment]
> but even now I can't compile, the error message is the same.

Okay that's my bad as I only copied the few LIBC headers needed from the ZIP file to Linux when I tested.

There's a Windows header [../../include/segment.inc](../blob/master/include/segment.inc#L8) and a Asmc header [./src/inc/segment.inc](../blob/master/source/asmc/src/inc/segment.inc#L8).

> ../../bin/asmc -nologo -nolib -Zp4 -elf -Cs -I../../include -Isrc/inc -Fo src/assemble.o src/assemble.asm

The order of include directories needs to be flipped there:

`../../bin/asmc -nologo -nolib -Zp4 -elf -Cs -Isrc/inc -I../../include`

**Comment by teoberi** [comment]
> Okay that's my bad as I only copied the few LIBC headers needed from the ZIP file to Linux when I tested.

@nidud be in love with Linux! 
Now yes, it can be compiled without the slightest error but only in a multilib environment.

**Comment by nidud** [comment]
> ASMC : fatal error A1901: Internal Assembler Error

There was a bug in the [longjmp()](https://github.com/open-watcom/open-watcom-v2/blob/master/bld/clib/startup/a/stjmp386.asm#L111) function from the Open Watcom Linux library. I added these functions to the source so this should be fixed in v2.22.22.

**Comment by teoberi** [comment]
> There was a bug in the longjmp() function from the Open Watcom Linux library. I added these functions to the source so this should be fixed in v2.22.22.

v2.22.22 -> 2.33.22

**Comment by darealshinji** [comment]
> If yes, is there any way to compile that 1st stage binary from source?

I wonder if it can be bootstrapped with another assembler?

**Comment by teoberi** [comment]
> I wonder if it can be bootstrapped with another assembler?

So depending on another assembler, I don't know if it's OK.

> An alternative might be to provide the assembled object files too.

So let's remove the executable (asmc) and have only the object files assembled. I don't know what to say yet, I think it still needs to be analyzed.

**Comment by darealshinji** [comment]
> > I wonder if it can be bootstrapped with another assembler?
> 
> So depending on another assembler, I don't know if it's OK.

Not as a requirement but as an option. This may be needed to get asmc added to some Linux distros so that the latest 7zip version can be compiled including assembler code. Regular users would just download the binaries provided here.



> > An alternative might be to provide the assembled object files too.
>
> So let's remove the executable (asmc) and have only the object files assembled. I don't know what to say yet, I think it still needs to be analyzed.

Not removing the executable but additionally providing assembled ELF object files. The idea would be that Linux distros who try to avoid binaries in source packages could only include the object files, link them during the build process and use that executable to assemble the .asm source and then check if the assembled object files are identical to the provided ones, you know, so that the source package includes assembled object files but the original corresponding source files are included as well. This _may_ be acceptable by some distros, I guess.

**Comment by darealshinji** [comment]
That would be a scenario for a very pedantic distro (like Debian?):
``` sh
# link bootstrap asmc
gcc -m32 -s -o asmc-bootstrap bootstrap/*.o

# assemble sources
./asmc-bootstrap -nologo -nolib -Zp4 -elf -Cs -Isrc/inc -I../../include -Fosrc/ src/*.asm

# link final asmc
gcc -m32 -s -o asmc src/*.o

# be pedantic and check if the new object files are identical to
# the ones used for bootstrapping

cd bootstrap
md5sum -b *.o > ../stage1.txt
cd ../src
md5sum -b *.o > ../stage2.txt
cd ..

if [ "x$(diff stage1.txt stage2.txt)" = "x" ]; then
  echo "ok"
else
  echo "error"
  exit 1
fi

# assemble and link asmc64
rm src/*.o
./asmc -D__ASMC64__ -nologo -nolib -Zp4 -elf -Cs -Isrc/inc -I../../include -Fosrc/ src/*.asm
gcc -m32 -s -o asmc64 src/*.o
```

**Comment by jloqfjgk** [comment]
> I wonder if it can be bootstrapped with another assembler?

It would be great if asmc can be bootstrapped with e.g. UASM (from [here](https://repology.org/project/uasm/versions) Alpine Linux already has package for UASM), but I don't know whether it is easy / trivial to make the bootstrap source UASM compatible...

**Comment by teoberi** [comment]
UASM is a mature product (it also has a pure 64-bit version) but its development seems to have problems from what I read somewhere, the developer doesn't have much time for it.
Also the compilation on Linux is broken, there are many pull requests to solve this but so far all without results.

**Comment by jloqfjgk** [comment]
I found a UASM fork which integrates the pull requests so that UASM can be compiled on Linux:
https://github.com/64kramsystem/UASM-linux

**Comment by darealshinji** [comment]
I also made a fork of UASM with fixes for Clang and GCC on Linux, if you're interrested:
https://github.com/darealshinji/UASM

**Comment by teoberi** [comment]
> I also made a fork of UASM with fixes for Clang and GCC on Linux, if you're interrested: https://github.com/darealshinji/UASM

I've already tried it, it compiles without any problems.
Ideally, the UASM maintainer should solve the problem using the patches suggested by the programmers who want to help.

**Comment by teoberi** [comment]
> This will also work:
> 
> ```
> ifdef @wordsize
> if @wordsize eq 8
>  x64 equ 1
> endif
> else
>  x64 equ 1
> endif
> ```

The solution was implemented by Igor Pavlov in 7-Zip 21.07 for Linux.
